File: 조성원/[LV3] 경주로 건설.js
const DIRECTIONS = [
  [0, 1],
  [1, 0],
  [0, -1],
  [-1, 0],
];

const isRange = (x, y, n) => x >= 0 && x < n && y >= 0 && y < n;
const isRoad = (x, y, board) => board[x][y] === 0;

function solution(board) {
  const n = board.length;

  const queue = [[0, 0, 0, -1]];
  const costs = Array.from({ length: n }, () =>
    Array(n).fill(Number.POSITIVE_INFINITY)
  );

  while (queue.length) {
    const [x, y, cost, dir] = queue.shift();

    if (costs[x][y] + 500 <= cost) continue;

    costs[x][y] = Math.min(costs[x][y], cost);

    for (const [i, [dx, dy]] of DIRECTIONS.entries()) {
      const nx = x + dx;
      const ny = y + dy;

      if (isRange(nx, ny, n) && isRoad(nx, ny, board)) {
        const newCost = cost + (dir === i || dir === -1 ? 100 : 600);
        queue.push([nx, ny, newCost, i]);
      }
    }
  }

  return costs[n - 1][n - 1];
}

File: 김현경/[LV3] 경주로 건설.py
from collections import deque

def solution(board):
    n = len(board)
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    cost = [[[float('inf')] * 4 for _ in range(n)] for _ in range(n)]
    queue = deque([(0, 0, -1, 0)])
    
    for i in range(4):
        cost[0][0][i] = 0
    
    while queue:
        x, y, prev_dir, cur_cost = queue.popleft()
        
        for new_dir, (dx, dy) in enumerate(directions):
            nx, ny = x + dx, y + dy
            
            if 0 <= nx < n and 0 <= ny < n and board[nx][ny] == 0:
                new_cost = cur_cost + (100 if prev_dir == new_dir or prev_dir == -1 else 600)
                if new_cost < cost[nx][ny][new_dir]:
                    cost[nx][ny][new_dir] = new_cost
                    queue.append((nx, ny, new_dir, new_cost))
                    
    return min(cost[n - 1][n - 1])